spark3/src/main/java/org/apache/iceberg/spark/procedures/BaseProcedure.java

```diff
@@ -75,25 +71,15 @@ protected BaseProcedure(TableCatalog tableCatalog) {
   }
 
   // we have to parse both namespace and name as they may be quoted
-  protected Identifier toIdentifier(String namespaceAsString, String name) {
-    String[] namespaceParts = parseMultipartIdentifier(namespaceAsString);
-
-    String[] nameParts = parseMultipartIdentifier(name);
-    Preconditions.checkArgument(nameParts.length == 1, "Name must consist of one part: %s", name);
-
-    return Identifier.of(namespaceParts, nameParts[0]);
-  }
-
-  private String[] parseMultipartIdentifier(String identifierAsString) {
+  protected Identifier toIdentifier(String identifier) {
+    Spark3Util.CatalogAndIdentifier catalogAndIdentifier;
     try {
-      ParserInterface parser = spark.sessionState().sqlParser();
-      Seq<String> namePartsSeq = parser.parseMultipartIdentifier(identifierAsString);
-      String[] nameParts = new String[namePartsSeq.size()];
-      namePartsSeq.copyToArray(nameParts);
-      return nameParts;
+      catalogAndIdentifier = Spark3Util.catalogAndIdentifier(spark, identifier, tableCatalog);
```

Comment:
Last time we talked, we thought about defaulting the catalog to the procedure catalog, not the default catalog.

I thought we would simply take the first parsed name part and check if `spark.sql.catalog._name_part_` is set. If yes, it means our identifier contains a catalog and we don't have to prepend the procedure catalog to name parts and should resolve the catalog and validate the resolved catalog matches tableCatalog. If there is no such catalog, we can just call `tableCatalog.load` with the constructed identifier.

Reply:
@RussellSpitzer @rdblue, thoughts?

Reply:
I did this based on some of the discussion me and Ryan were having

```
We know we have a catalog if there is a registered catalog with the name





11:45
In Spark, the rules are:
1. If the identifier is a single part, it is a catalog name. Fill in current catalog and current namespace
2. If the first part of the identifier is a known catalog, it is a fully qualified name
3. If the first part is not a known catalog, fill in the current catalog (not namespace because it is already there)
11:46
The only difference here would be that we consider the current catalog to be the procedure catalog because this is in the procedure's context
```

I thought this was reasonable but we could force the catalog as well

Reply:
Isn't that what `Spark3Util.catalogAndIdentifier` does? I think the only difference is that it uses Spark's catalogManager instead of simply checking table properties.

Reply:
I think I missed an update here.

Reply:
I added an argument to catalogAndIdentifier, the last parameter is the "current catalog" that's used